vwifi.c

```diff
@@ -403,24 +430,163 @@ static struct cfg80211_ops owl_cfg_ops = {
     .get_station = owl_get_station,
 };
 
+static int owl_ndo_open(struct net_device *dev)
+{
+    /* The first byte is '\0' to avoid being a multicast
+     * address (the first byte of multicast addrs is odd).
+     */
+    memcpy(dev->dev_addr, "\0owl0", ETH_ALEN);
+    struct owl_context *owl = ndev_get_owl_context(dev)->owl;
+    INIT_LIST_HEAD(&owl->rx_queue);
+    netif_start_queue(dev);
+    return 0;
+}
+
+static int owl_ndo_stop(struct net_device *dev)
+{
+    struct owl_context *owl = ndev_get_owl_context(dev)->owl;
+    struct owl_packet *pkt, *is = NULL;
+    list_for_each_entry_safe (pkt, is, &owl->rx_queue, list) {
+        list_del(&pkt->list);
+        kfree(pkt);
+    }
+    netif_stop_queue(dev);
+    return 0;
+}
+
+static struct net_device_stats *owl_ndo_get_stats(struct net_device *dev)
+{
+    struct owl_context *owl = ndev_get_owl_context(dev)->owl;
+    return &owl->stats;
+}
+
+/*
+ * Receive a packet: retrieve, encapsulate and pass over to upper levels
+ */
+static void owl_rx(struct net_device *dev)
+{
+    struct sk_buff *skb;
+    int hit_ours_pkt = 0;
+    int offset = 0;
+    struct owl_context *owl = ndev_get_owl_context(dev)->owl;
+    struct owl_packet *pkt;
+    if (list_empty(&owl->rx_queue)) {
+        printk(KERN_NOTICE "owl rx: No packet in rx_queue\n");
+        goto out;
+    }
+    pkt = list_first_entry(&owl->rx_queue, struct owl_packet, list);
+    /* TODO: Modify the packet and forward to destination */
+    struct ethhdr *eth = (struct ethhdr *) &pkt->data[offset];
+    offset += sizeof(struct ethhdr);
+    /* Process ARP Request */
+    if (eth->h_proto == htons(ETH_P_ARP)) {
+        printk(KERN_NOTICE "ARP request received, send ARP reply to sta\n");
+        hit_ours_pkt = 1;
+        // dst mac
+        ether_addr_copy(eth->h_dest, dev->dev_addr);
+        // src mac
+        ether_addr_copy(eth->h_source, owl->connecting_bssid);
+        // arp reply opcode 0x2
+        struct arphdr *arp = (struct arphdr *) &pkt->data[offset];
+        offset += sizeof(struct arphdr);
+        arp->ar_op = htons(ARPOP_REPLY);
+        // TODO: Need more efficient to get packet field offset
+        // Update Sender
+        u8 tmp[10];
+        memcpy(tmp, &pkt->data[22], 10);
+        ether_addr_copy(&pkt->data[32],
+                        eth->h_source);  // Sender mac is station source mac.
+        memcpy(&pkt->data[22], &pkt->data[32], 10);
+        // Update Target
+        memcpy(&pkt->data[32], tmp, 10);
+    }
+    /* Process ICMP Request */
```

Comment:
I am not confident if this function is the proper place where ICMP manipulation should locate or not.

Reply:
If you proceed ICMP requests in this function, where should we implement further TCP/UDP support?

Reply:
Doing the protocol stuff in `owl_rx` is not a good method. It's is better to let the kernel network stack handle the protocol. I could try to use  namespace or veth.

Reply:
Good idea. Look forward to the use of namespace.